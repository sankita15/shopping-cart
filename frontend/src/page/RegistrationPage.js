import React from 'react';
import {
    Label,
    InputGroup,
    InputGroupAddon,
    Input,
    Button,
    Col,
    Row,
    Card,
    CardBody,
    FormFeedback,
    FormGroup,
} from 'reactstrap';

export default class RegistrationPage extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            firstName: null,
            lastName: null,
            username: null,
            email: null,
            password: null,
            confirmPasswordMatch: null,
        };
    }

    handleFieldChange(fieldName, fieldValue, regex) {
        const result = regex.test(fieldValue) ? fieldValue : '';
        this.setState({
            [fieldName]: result,
        });
    }

    confirmPassword(fieldValue) {
        const { password } = this.state;
        this.setState({
            confirmPasswordMatch: password === fieldValue,
        });
    }

    canSubmit() {
        const { firstName, lastName, username, email, password, confirmPasswordMatch } = this.state;
        return firstName && lastName && username && email && password && confirmPasswordMatch;
    }

    _handleSubmit() {
        const { email, username, password, firstName, lastName } = this.state;

        return fetch('/api/users/signup', {
            headers: {
                'Content-Type': 'application/json',
            },
            method: 'POST',
            body: JSON.stringify({
                email,
                username,
                password,
                firstName,
                lastName,
            }),
        })
            .then(res => (res.ok ? Promise.resolve() : Promise.reject(res.status)))
            .catch(status => console.warn('Sign up failed', status));
    }

    async submitForm() {
        await this._handleSubmit();
        this.navigateToPage('/login');
    }

    navigateToPage = route => window.location.assign(route);

    render() {
        const { email, username, password, confirmPasswordMatch, firstName, lastName } = this.state;

        return (
            <div className="registration-page">
                <Col>
                    <Row sm={3}>
                        <Card>
                            <CardBody>
                                <Label className="heading1">Signup with ONE-STOP-SHOP</Label>
                                <FormGroup>
                                    <Input
                                        className="firstname"
                                        placeholder="Enter Your Firstname"
                                        onChange={e => this.handleFieldChange('firstName', e.target.value, /^[a-zA-Z0-9]{3,20}$/)}
                                        invalid={!firstName && firstName !== null}
                                    />
                                    <FormFeedback>
                                            firstName cannot be less than 3 letters
                                    </FormFeedback>
                                </FormGroup>
                                <FormGroup>
                                    <Input
                                        className="lastname"
                                        placeholder="Enter your Lastname"
                                        onChange={e => this.handleFieldChange('lastName', e.target.value, /^[a-zA-Z0-9]{3,25}$/)}
                                        invalid={!lastName && lastName !== null}
                                    />
                                    <FormFeedback>
                                            Lastname cannot be less than 3 letters
                                    </FormFeedback>
                                </FormGroup>
                                <InputGroup className="email">
                                    <InputGroupAddon addonType="prepend">@</InputGroupAddon>
                                    <Input
                                        placeholder="Your Email address"
                                        onChange={e => this.handleFieldChange('email', e.target.value, /^\w+@[a-zA-Z_]+?\.[a-zA-Z]{2,3}$/)}
                                        invalid={!email && email !== null}
                                    />
                                    <FormFeedback>
                                        Please provide valid email address.
                                        Email must include @ followed by hostname.
                                    </FormFeedback>
                                </InputGroup>
                                <FormGroup>
                                    <br />
                                    <Input
                                        className="username"
                                        placeholder="Choose Your Username"
                                        onChange={e => this.handleFieldChange('username', e.target.value, /^[a-zA-Z0-9]{3,10}$/)}
                                        invalid={!username && username !== null}
                                    />
                                    <FormFeedback>
                                    Please choose a username of length between 3 to 10
                                    </FormFeedback>
                                </FormGroup>
                                <FormGroup>
                                    <Input
                                        className="password"
                                        placeholder="Choose Your Password"
                                        type="password"
                                        onChange={e => this.handleFieldChange('password', e.target.value, /^[a-zA-Z0-9!@#$%^&*]{4,8}$/)}
                                        invalid={!password && password !== null}
                                    />
                                    <FormFeedback>
                                    Please choose a password of length between 4 to 8.
                                    </FormFeedback>
                                </FormGroup>
                                <FormGroup>
                                    <Input
                                        className="password"
                                        placeholder="Confirm Your Password"
                                        type="password"
                                        onChange={e => this.confirmPassword(e.target.value)}
                                        invalid={!confirmPasswordMatch && confirmPasswordMatch !== null}
                                    />
                                    <FormFeedback>
                                        Password & Confirm Password does not match.
                                    </FormFeedback>
                                </FormGroup>
                                <Button type="submit" disabled={!this.canSubmit()} className="register-button" onClick={() => this.submitForm()}>Register</Button>
                                <Row className="login-back-page"><a className="login-link" href="/login">Back To Login</a></Row>
                            </CardBody>
                        </Card>
                    </Row>
                </Col>
            </div>
        );
    }
}
