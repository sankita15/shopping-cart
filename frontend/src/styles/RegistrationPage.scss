
.registration{
  margin-left: 496px;
  margin-top: 183px;
}
.email{
  width: 387px
}
.username{
  width: 387px
}
.password{
  width: 387px
}
.firstname{
  width: 387px
}
.lastname{
  width: 387px
}

.register-button{
  width: 387px;
  margin-top: 50px;
  background-color: #007bff;
}
.registration-page{
  margin-left: 490px;
  margin-top: 103px;
}
.heading1{
  font-size: 27px;
  margin-bottom: 61px;
  margin-left: 11px;
  font-style: italic;
}
.login-back-page{
  margin-left: 286px;
  margin-top: 15px;
}
