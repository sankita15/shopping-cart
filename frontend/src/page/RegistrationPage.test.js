import React from 'react';
import { shallow } from 'enzyme';
import { Label, Card, Button, CardBody, InputGroup, Input } from 'reactstrap';
import RegistrationPage from './RegistrationPage';


describe('Registration Page', () => {
    const INVALID_EMAIL = 'xyz@abc';
    const VALID_EMAIL = 'xyz@abc.com';

    const INVALID_FIRSTNAME = 'ab'; // less than 3 & more than 10
    const VALID_FIRSTNAME = 'abcdef';

    const INVALID_LASTNAME = 'ab'; // less than 3 & more than 10
    const VALID_LASTNAME = 'abcdef';

    const INVALID_USERNAME = 'ab'; // less than 3 & more than 10
    const VALID_USERNAME = 'abcdef';

    const INVALID_PASSWORD = 'abcdefghij'; // less than 4 & more than 8
    const VALID_PASSWORD = 'toto';

    const SIGNUP_BODY = {
        email: VALID_EMAIL,
        username: VALID_USERNAME,
        password: VALID_PASSWORD,
        firstName: VALID_FIRSTNAME,
        lastName: VALID_LASTNAME,
    };

    window.location.assign = jest.fn();

    let component;
    let firstName;
    let lastName;
    let email;
    let username;
    let password;
    let confirmPassword;
    let button;
    let heading;
    let card;

    const createWrapper = () => shallow(<RegistrationPage />);

    beforeEach(() => {
        component = createWrapper();

        heading = component.find(Label);
        card = component.find(Card);
        firstName = component.find(Card).find(CardBody).find(Input).at(0);
        lastName = component.find(Card).find(CardBody).find(Input).at(1);
        email = component.find(Card).find(CardBody).find(InputGroup).find(Input);
        username = component.find(Card).find(CardBody).find(Input).at(3);
        password = component.find(Card).find(CardBody).find(Input).at(4);
        confirmPassword = component.find(Card).find(CardBody).find(Input).at(5);
        button = component.find(Button);
    });

    const testSetUpMethod = () => {
        firstName.props().onChange({ target: { value: VALID_FIRSTNAME } });
        lastName.props().onChange({ target: { value: VALID_LASTNAME } });
        username.props().onChange({ target: { value: VALID_USERNAME } });
        password.props().onChange({ target: { value: VALID_PASSWORD } });
        confirmPassword.props().onChange({ target: { value: VALID_PASSWORD } });
    };

    afterEach(() => {
        fetch.resetMocks();
    });

    it('should match snapshot', () => {
        expect(component).toMatchSnapshot();
    });

    it('should render Registration Page field', () => {
        expect(heading.length).toEqual(1);
        expect(card.length).toEqual(1);
        expect(button.length).toEqual(1);
    });

    it('should render invalid field when email address is not valid', () => {
        expect(email.prop('invalid')).toBeFalsy();

        email.props().onChange({ target: { value: INVALID_EMAIL } });

        const email1 = component.find(Card).find(CardBody).find(InputGroup).find(Input);

        expect(email1.prop('invalid')).toBeTruthy();
    });

    it('should render invalid field when username is not valid', () => {
        expect(username.prop('invalid')).toBeFalsy();

        username.props().onChange({ target: { value: INVALID_USERNAME } });

        const username1 = component.find(Card).find(CardBody).find(Input).at(3);

        expect(username1.prop('invalid')).toBeTruthy();
    });

    it('should render invalid field when password is not valid', () => {
        expect(password.prop('invalid')).toBeFalsy();

        password.props().onChange({ target: { value: INVALID_PASSWORD } });

        const password1 = component.find(Card).find(CardBody).find(Input).at(4);

        expect(password1.prop('invalid')).toBeTruthy();
    });

    it('should match password with confirm password field', () => {
        password.props().onChange({ target: { value: VALID_PASSWORD } });
        confirmPassword.props().onChange({ target: { value: INVALID_PASSWORD } });

        const confirmPassword1 = component.find(Card).find(CardBody).find(Input).at(5);

        expect(confirmPassword1.prop('invalid')).toBeTruthy();
    });

    it('should render invalid field when firstName is not valid', () => {
        expect(firstName.prop('invalid')).toBeFalsy();

        firstName.props().onChange({ target: { value: INVALID_FIRSTNAME } });

        const firstName1 = component.find(Card).find(CardBody).find(Input).at(0);

        expect(firstName1.prop('invalid')).toBeTruthy();
    });

    it('should render invalid field when lastName is not valid', () => {
        expect(lastName.prop('invalid')).toBeFalsy();

        lastName.props().onChange({ target: { value: INVALID_LASTNAME } });

        const lastName1 = component.find(Card).find(CardBody).find(Input).at(1);

        expect(lastName1.prop('invalid')).toBeTruthy();
    });

    it('should disable button if any field is invalid', () => {
        testSetUpMethod();

        email.props().onChange({ target: { value: INVALID_EMAIL } });

        expect(button.props().disabled).toBeTruthy();
    });

    it('should enable button if all the fields are valid', () => {
        expect(button.props().disabled).toBeTruthy();

        testSetUpMethod();
        email.props().onChange({ target: { value: VALID_EMAIL } });

        component.update();

        expect(component.find(Button).props().disabled).toBeFalsy();
    });

    it('should call fetch for POST call for sign up', async () => {
        fetch.mockResolvedValueOnce({ json: () => Promise.resolve(), ok: true });

        testSetUpMethod();
        email.props().onChange({ target: { value: VALID_EMAIL } });

        button.simulate('click');

        component.update();

        // eslint-disable-next-line no-undef
        await flushPromises();

        expect(fetch).toHaveBeenCalledWith('/api/users/signup', {
            headers: {
                'Content-Type': 'application/json',
            },
            method: 'POST',
            body: JSON.stringify(SIGNUP_BODY),
        });

        expect(window.location.assign).toHaveBeenCalledWith('/login');
    });
});
