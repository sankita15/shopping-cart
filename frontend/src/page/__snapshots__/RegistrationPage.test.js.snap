// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registration Page should match snapshot 1`] = `
<div
  className="registration-page"
>
  <Col
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <Row
      sm={3}
      tag="div"
    >
      <Card
        tag="div"
      >
        <CardBody
          tag="div"
        >
          <Label
            className="heading1"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            Signup with ONE-STOP-SHOP
          </Label>
          <FormGroup
            tag="div"
          >
            <Input
              className="firstname"
              invalid={false}
              onChange={[Function]}
              placeholder="Enter Your Firstname"
              type="text"
            />
            <FormFeedback
              tag="div"
            >
              firstName cannot be less than 3 letters
            </FormFeedback>
          </FormGroup>
          <FormGroup
            tag="div"
          >
            <Input
              className="lastname"
              invalid={false}
              onChange={[Function]}
              placeholder="Enter your Lastname"
              type="text"
            />
            <FormFeedback
              tag="div"
            >
              Lastname cannot be less than 3 letters
            </FormFeedback>
          </FormGroup>
          <InputGroup
            className="email"
            tag="div"
          >
            <InputGroupAddon
              addonType="prepend"
              tag="div"
            >
              @
            </InputGroupAddon>
            <Input
              invalid={false}
              onChange={[Function]}
              placeholder="Your Email address"
              type="text"
            />
            <FormFeedback
              tag="div"
            >
              Please provide valid email address. Email must include @ followed by hostname.
            </FormFeedback>
          </InputGroup>
          <FormGroup
            tag="div"
          >
            <br />
            <Input
              className="username"
              invalid={false}
              onChange={[Function]}
              placeholder="Choose Your Username"
              type="text"
            />
            <FormFeedback
              tag="div"
            >
              Please choose a username of length between 3 to 10
            </FormFeedback>
          </FormGroup>
          <FormGroup
            tag="div"
          >
            <Input
              className="password"
              invalid={false}
              onChange={[Function]}
              placeholder="Choose Your Password"
              type="password"
            />
            <FormFeedback
              tag="div"
            >
              Please choose a password of length between 4 to 8.
            </FormFeedback>
          </FormGroup>
          <FormGroup
            tag="div"
          >
            <Input
              className="password"
              invalid={false}
              onChange={[Function]}
              placeholder="Confirm Your Password"
              type="password"
            />
            <FormFeedback
              tag="div"
            >
              Password & Confirm Password does not match.
            </FormFeedback>
          </FormGroup>
          <Button
            className="register-button"
            color="secondary"
            disabled={true}
            onClick={[Function]}
            tag="button"
            type="submit"
          >
            Register
          </Button>
          <Row
            className="login-back-page"
            tag="div"
          >
            <a
              className="login-link"
              href="/login"
            >
              Back To Login
            </a>
          </Row>
        </CardBody>
      </Card>
    </Row>
  </Col>
</div>
`;
